### PR TITLE
Add Bayesian Synthetic Difference-in-Differences

### PR DESCRIPTION
**Pull request by thomaspinder**
Implements the cut-posterior SDiD formulation of SDiD as a new experiment class. Unit and time weights are estimated jointly in a single `SyntheticDifferenceInDifferencesWeightFitter` model, and the ATT is computed analytically via the double-difference formula. Includes the California Proposition 99 dataset, a Jupyter notebook demo, and full test coverage.

Depends on `feat/softmax-weighted-sum-fitter` (PR #822) for the `_softmax_simplex_weights` helper and `SoftmaxWeightedSumFitter` base.

Relevant Issue: #47

### CODE REVIEW

**Comment by juanitorduz** [comment]
So cool! 💪 . Is this one ready for review?

**Comment by thomaspinder** [comment]
> So cool! 💪 . Is this one ready for review?

It is!

**Comment by thomaspinder** [comment]
pre-commit.ci run

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+
+from causalpy.custom_exceptions import BadIndexException
+from causalpy.date_utils import _combine_datetime_indices, format_date_axes
+from causalpy.plot_utils import plot_xY
+from causalpy.pymc_models import PyMCModel, SyntheticDifferenceInDifferencesWeightFitter
+from causalpy.reporting import EffectSummary
+
+from .base import BaseExperiment
+
+LEGEND_FONT_SIZE = 12
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:36]: Should this live in a config file? and not in this module?

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+        y = {"unit": y_unit, "time": y_time}
+
+        COORDS = {
+            "coeffs": self.control_units,
+            "obs_ind": np.arange(T_pre),
+            "coeffs_raw": self.control_units[1:],
+            "obs_ind_raw": list(range(1, T_pre)),
+        }
+
+        self.model.fit(X=X, y=y, coords=COORDS)
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:276]: Shall we allow the user to pass kwargs for the fit method? Say I wanna sample with Nutpie and add a random seed?

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+        assert self.model.idata is not None, "Model fitting failed to produce idata"
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:277]: [Opinion]: Avoid assert statements, see https://docs.astral.sh/ruff/rules/assert/

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+
+        # ---- Extract weight posteriors ----
+        omega = self.model.idata.posterior["omega"].values  # (chain, draw, N_co)
+        lam = self.model.idata.posterior["lam"].values  # (chain, draw, T_pre)
+        omega0 = self.model.idata.posterior["omega0"].values  # (chain, draw)
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:282]: .to_numpy() instead of .values ?

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+        """
+        if isinstance(self.model, RegressorMixin):
+            raise NotImplementedError(
+                "OLS estimation for SyntheticDifferenceInDifferences is not yet "
+                "implemented. Please use a PyMC model."
+            )
+
+        # Full panel data as numpy
+        Y_co = self.data[self.control_units].values.T  # (N_co, T)
+        y_tr = self.data[self.treated_units].values.mean(
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:226]: .to_numpy() instead of .values ?

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+        self.datapost_treated = xr.DataArray(
+            self.datapost[self.treated_units],
+            dims=["obs_ind", "treated_units"],
+            coords={
+                "obs_ind": self.datapost[self.treated_units].index,
+                "treated_units": self.treated_units,
+            },
+        )
+
+    def algorithm(self) -> None:
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:206]: Suggestion: this function is well documented, but it's rather long. Maybe we can extract the steps in private functions and then the `algorithm` just calls the steps? (This way, we can unit-test the individual steps

File: causalpy/experiments/synthetic_difference_in_differences.py [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+                "treated_units": [self.treated_units[0]],
+            },
+        )
+        ds = xr.Dataset({"mu": mu_da})
+        return az.InferenceData(posterior_predictive=ds)
+
+    def summary(self, round_to: int | None = None) -> None:
+        """Print summary of main results.
+
+        :param round_to:
```
Comment (juanitorduz) [inline causalpy/experiments/synthetic_difference_in_differences.py:424]: We should use Numpy- style docstrings ;)

**Review by juanitorduz — COMMENTED** [review]
This looks great! just left some initial comments :)

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 2 above)
Comment (juanitorduz) [reply causalpy/experiments/synthetic_difference_in_differences.py:276]: ahh ok! I now understand the API from the notebook

```python
result = cp.SyntheticDifferenceInDifferences(
    df,
    treatment_time,
    control_units=control_units,
    treated_units=treated_units,
    model=cp.pymc_models.SyntheticDifferenceInDifferencesWeightFitter(
        sample_kwargs={
            "target_accept": 0.95,
            "random_seed": seed,
            "tune": 3000,
            "draws": 2000,
        },
        priors={
            "sigma_omega": Prior("HalfNormal", sigma=y_sd),
            "sigma_lambda": Prior("HalfNormal", sigma=y_sd),
            "omega0": Prior("Normal", mu=0, sigma=y_sd * 2),
            "lambda0": Prior("Normal", mu=0, sigma=y_sd * 2),
        },
    ),
)
```

File: docs/source/notebooks/sdid_pymc.ipynb [hunk 8]
```diff
@@ -0,0 +1,864 @@
+{
```
Comment (juanitorduz) [inline docs/source/notebooks/sdid_pymc.ipynb:1]: <p>maybe remove the "---" in the text and use "," for example? (llms over do the --- imo XD)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/823/discussion/#comment-3092266632'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

Comment (juanitorduz) [inline docs/source/notebooks/sdid_pymc.ipynb:1]: <p>maybe we could use az.plot_forest() (just a suggestion as parplots with vertical lines are not that easy to read for me) <a href="https://python.arviz.org/projects/plots/en/stable/api/generated/arviz_plots.plot_forest.html" target="_blank">https://python.arviz.org/projects/plots/en/stable/api/generated/arviz_plots.plot_forest.html</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/823/discussion/#comment-3092266636'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

Comment (juanitorduz) [inline docs/source/notebooks/sdid_pymc.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #22.        #     }
</pre><p>remove commented code?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/CausalPy/pull/823/discussion/#comment-3092266643'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='16'/>

**Comment by juanitorduz** [comment]
@thomaspinder this looks great! left some minor suggestions, let me know what you think about it :)

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 1 above)
Comment (thomaspinder) [reply causalpy/experiments/synthetic_difference_in_differences.py]: Tbh, this is somewhat superfluous. Removed.

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 2 above)
Comment (thomaspinder) [reply causalpy/experiments/synthetic_difference_in_differences.py]: Precisely :) Are you ok with this API, or does your original comment still stand and you're in favour of a `**kwargs` style argument in the SDiD model? Generally, my principle was minimising the number of places to inject this type of information, and keeping the number of `**kwargs` to a minimum.

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 3 above)
Comment (thomaspinder) [reply causalpy/experiments/synthetic_difference_in_differences.py]: Cool! Yes, this is indeed a good point. I've added an `AttributeError`, but let me know if you'd prefer some type of custom error handler.

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 4 above)
Comment (thomaspinder) [reply causalpy/experiments/synthetic_difference_in_differences.py]: Done. Curiosity Q - Is this preferable for any reason beyond explicitness?

File: docs/source/notebooks/sdid_pymc.ipynb (hunk 8 above)
Comment (thomaspinder) [reply docs/source/notebooks/sdid_pymc.ipynb:1]: Yes, this is fair, and an oversight. The notebook was mainly an LLM conversion of my blog post XD

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 6 above)
Comment (thomaspinder) [reply causalpy/experiments/synthetic_difference_in_differences.py:207]: Yeah this is a good point. Addressed.

File: docs/source/notebooks/sdid_pymc.ipynb (hunk 8 above)
Comment (thomaspinder) [reply docs/source/notebooks/sdid_pymc.ipynb:1]: I'm 50-50. I sometimes find forest plots awkward to read; however, it's probably more in keeping with the PyMC/CausalPy docs.

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 2 above)
Comment (juanitorduz) [reply causalpy/experiments/synthetic_difference_in_differences.py]: This makes sense (and it is documented in the example as well :) ), maybe @drbenvincent has an opinion here.

File: causalpy/experiments/synthetic_difference_in_differences.py (hunk 4 above)
Comment (juanitorduz) [reply causalpy/experiments/synthetic_difference_in_differences.py]: because of https://docs.astral.sh/ruff/rules/pandas-use-of-dot-values/ (we might need to add it into the linter)

**Review by juanitorduz — APPROVED** [review]
great! thank you @thomaspinder ! It seems tests failures are because of other reasons, right? cc @drbenvincent

**Comment by thomaspinder** [comment]
Hey @drbenvincent, just want to check if there's any movement here re. failing upstream tests? Let me know if there's anything I should do before we're ready to merge.

**Comment by drbenvincent** [comment]
Planning on looking at this today :) Sorry it's taken a while
